src/dns.rs

```diff
@@ -0,0 +1,464 @@
+use crate::topology::config::GlobalOptions;
+use futures::{Async, Future, Poll};
+use hyper::client::connect::dns::{Name, Resolve};
+use std::io;
+use std::net::ToSocketAddrs;
+use std::net::*;
+use std::vec::IntoIter;
+use trust_dns_resolver::config::*;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_resolver::AsyncResolver;
+use trust_dns_resolver::BackgroundLookupIp;
+
+/// Default port for DNS service.
+const DNS_PORT: u16 = 53;
+
+/// Default version will behave identically to std::net::ToSocketAddrs implementations.
+#[derive(Clone, Default)]
+pub struct DnsResolver(Option<AsyncResolver>);
+
+impl DnsResolver {
+    /// (Self,errors)
+    pub fn new(config: &GlobalOptions) -> Result<Self, String> {
+        if config.dns.is_empty() {
+            return Ok(DnsResolver::default());
+        }
+
+        let mut resolve_config = ResolverConfig::new();
+
+        let mut errors = Vec::new();
+
+        for s in config.dns.iter() {
+            let parsed = s
+                .parse::<SocketAddr>()
+                .or_else(|_| s.parse::<IpAddr>().map(|ip| SocketAddr::new(ip, DNS_PORT)));
+            match parsed {
+                Ok(socket_addr) => resolve_config.add_name_server(NameServerConfig {
+                    socket_addr,
+                    protocol: Protocol::Udp,
+                    tls_dns_name: None,
+                }),
+                Err(error) => errors.push(format!(
+                    "DNS server IP {:?} is invalid. Parse Error: {:?}",
+                    s, error
+                )),
+            }
+        }
+
+        if !errors.is_empty() {
+            return Err(format!(
+                "DNS resolver initialization failed with errors: {:?}",
+                errors
+            ));
+        }
+
+        let mut options = ResolverOpts::default();
+        options.attempts = 2;
+        options.validate = false;
+        options.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+        options.use_hosts_file = false;
```

Comment:
what is the reason to not let the host file?

Reply:
Because the configured resolver is meant to query configured DNS servers, and only if that fails to use host's resolver.

Reply:
I don't think this is accurate, disabling this would mean that the hosts file will never be used. Have you seen different behavior?

Reply:
You are true, but when I said 

>  and only if that fails to use host's resolver/config.

I meant that at that point we will deliberately use host resolver using Rusts implementation of `ToSocketAddrs` trait. Not that `trust_dns_resolver` would do it, as I presume you have interpreted it.

Reply:
Yes, you're correct. Though, I would say that is different from how I believe most client resolvers work, i.e. I believe the hosts file is referenced before querying DNS, just FYI.

Reply:
Interesting, but I think having custom DNS servers at the front will be a simpler model for the user. Especially for those who aren't even aware of that file.

Reply:
In my experience, and of course this may differ for others, hosts is generally empty (beside local host) and is only used by people who expect it to override any DNS resolution, ie before a query is made.

I’ve seen it used for local integration tests a lot for various pieces of software. 

Of course I’m not saying you shouldn’t do what you want here, but I’m just letting you know that this will behave differently from standard resolvers.

Reply:
I agree with @bluejekyll the host file should be checked first. Its an override point. Custom DNS servers just allow a user to augment what they may already have not completely replace.

Reply:
Those are some good reason. Alright I'll change it.

@bluejekyll thanks for sticking to it.